scripts/azure-pipelines/utils/AzureResourceUtils.psm1

```diff
@@ -446,6 +446,80 @@ function Remove-AzPrivateEndpointConnectionInResourceGroup {
   }
 }
 
+function Remove-AzKeyVaultInResourceGroup {
+  [CmdletBinding()]
+  param (
+    [Parameter(mandatory = $true)]
+    [string]$ResourceGroupName
+  )
+  
+  $keyVaults = Get-AzKeyVault -ResourceGroupName $ResourceGroupName
+  
+  foreach ($keyVault in $keyVaults) {
+    Write-Host "Removing Key vault" $keyVault.VaultName "..."
+    Remove-AzKeyVault -VaultName $keyVault.VaultName -ResourceGroupName $ResourceGroupName -Force
+
+    if (-not $keyVault.EnablePurgeProtection) {
```

Comment:
if enabled, would it maybe be a good idea to return a warning? For example:

```suggestion
    if ($keyVault.EnablePurgeProtection) {
       Write-Warning ('Key vault [{0}] had purge protection enabled. The retention time is [{1}]. Please wait until after this period before re-running the test.' -f  $keyVault.VaultName, $keyVault.SoftDeleteRetentionInDays)
    }
    else {
```

Alternatively, you could also move the purge outside the parent loop, fetch all purged key vaults via `Get-AzKeyVault -InRemovedState` and iterate over them to print the date when they are deleted (as it is not available otherwise):

```PowerShell
$softDeletedKeyVaults = Get-AzKeyVault -InRemovedState | Where-Object { $_.VaultName -in $keyVaults.VaultName }

foreach($softDeletedKeyVault in $softDeletedKeyVaults) {
  if (-not ($keyVaults | Where-Object { $_.VaultName -eq $softDeletedKeyVault.VaultName}).EnablePurgeProtection) {
      Write-Host "Purging Key vault" $softDeletedKeyVault.VaultName "..."
      Remove-AzKeyVault -VaultName $softDeletedKeyVault.VaultName -ResourceGroupName $ResourceGroupName -InRemovedState -Force
    } else {
       Write-Warning ('Key vault [{0}] had purge protection enabled. It will automatically be purged by [{1}]' -f  $softDeletedKeyVault .VaultName, $softDeletedKeyVault .ScheduledPurgeDate)
    }
}
```

Reply:
Good suggestion! Added a warning for purge protection enabled vaults. The alternative solution also reminds me that we cannot GET the soft delete enabled resources once the test resource group is removed, so I changed `Remove-TestResourceGroup.ps1` to purge those resources before removing the resource group.